### DISCUSSION TITLE
What is the best way to write huge tables to an XLSX?

### DISCUSSION THREAD

**Discussion by MorgeMoensch** [ClosedXML/ClosedXML#1986] [Q&A]
I have huge, dynamic database tables that I need to export to XLSX files. We're talking about multiple 100k, up to low millions of rows (although the latter is rare).
Previously, we simply generated CSVs which the users could then open in Excel. But it's not possible to lable CSVs with AIP (Azure Information Protection) - which is a business requirement, so generating CSVs is no longer an option.
With ClosedXML, the `saveAs` is the biggest pain point. With 200k rows and 34 columns, `.insertData()` takes 7'500ms, and `saveAs()` 39'000ms. 

Do you know of a good way to write those excels? Maybe it's possible to convert a CSV to an XLSX?

I'm opening the discussion because this was suggested by @vbjay. Thanks for any hints or ideas.

**Comment by igitur** [comment]
I haven't done this in many year, but long ago, I set up a basic Excel template file, with the column headers prepopulated (table per sheet) and used ADO.NET to connect to the .xlsx file and write out the records. You can find the connection string at https://www.connectionstrings.com/excel/ . I can't remember what number of rows I exported or what the performance was, but it worked. Something like this: https://www.aspsnippets.com/Articles/Export-Data-to-Excel-Sheet-using-ADO.Net-and-C.aspx

**Reply by MorgeMoensch** [reply]
Thanks for your response. 

Maybe I need to specify that my application is running on a (Linux) Server and the user can download XLSX files. Would you therefore recommend running a headless-excel on the server? I would have to look at it, whether that's even possible.
